### PR TITLE
Identity v3: Users Get / List

### PR DESCRIPTION
**Pull request by jtopjian**
For #287

### CODE REVIEW

**Comment by jtopjian** [comment]
@jrperritt Ready for review

**Comment by jtopjian** [comment]
Found a missing field while working on Create. That's all of them now.

**Comment by jtopjian** [comment]
Nope. Something is strange here. Making WIP for now.

**Comment by jtopjian** [comment]
@jrperritt Looks like Keystone also supports an arbitrary number of properties like Glance does in #260. There's little documentation on this, but search for "email" [here](https://github.com/openstack/keystone/blob/464ec0b3da48d0ca294dbf58e7409d013a2c4528/doc/source/devref/api_curl_examples.rst) and note this [`to_dict`](https://github.com/openstack/keystone/blob/a926700cf49895313f553b37a9186c625fc9ce4d/keystone/common/sql/core.py#L138) function. 

The second link is noteworthy because it seems Keystone switches between adding the extra properties to an `extra` dict _or_ just adds them to the general top-level JSON structure.

So while I was trying to avoid #260, I ended up having to solve it anyway.

No doubt the solution I came up with in the latest commit isn't the cleanest way of doing this. Can you take a look when you have time?

**Comment by jtopjian** [comment]
@jrperritt The `extra` field here is what I was talking about for #194. In [this](https://github.com/gophercloud/gophercloud/pull/288/commits/8c6fd3a301f410fe58fcfa5939e509b40928454b) commit, I'm using `reflect` to bundle the extra fields into the `extra` field. I'm all thumbs when it comes to `reflect`, so let me know if there's a better way of doing this.

In the case of this PR, `extra` is the name of the actual field. #316 added `AdditionalProperties` to the request body because that was also the name in the API. #260 will be `Properties`. But when there are cases when we add our own "extra" field, we should probably standardize on a name. `Extra`?

**Comment by jrperritt** [comment]
>  let me know if there's a better way of doing this.

I think that way is correct. That functionality would seem appropriate to put in a utility file in the `internal` package for other `Extract` function to use.

> added AdditionalProperties to the request body because that was also the name in the API

Ah, I didn't realize that.

> we should probably standardize on a name. Extra?

Other ideas: `Overflow`, `More`, `Other`.

**Comment by jrperritt** [comment]
@jtopjian Was this just waiting on `RemainingKeys`?

**Comment by jtopjian** [comment]
@jrperritt Yeah, I'll be working on this this week.

**Comment by jtopjian** [comment]
@jrperritt This is ready for review.

I'm not sure what happened, but somewhere along the way my branch got mangled. Since I had to rebase, I figure I'd squash the commits.

I reviewed everything today and even found a new user attribute of `Options`. The links in #287 have been updated, too.

File: openstack/identity/v3/users/requests.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+// ListOpts allows you to query the List method.
+type ListOpts struct {
+	// DomainID filters the response by a domain ID.
+	DomainID string `q:"domain_id"`
+
+	// Enabled filters the response by enabled users.
+	Enabled *bool `q:"enabled"`
+
+	// IdpID filters the response by an Identity Provider ID.
+	IdPID string `q:"enabled"`
```
Comment (jrperritt) [inline openstack/identity/v3/users/requests.go]: `q:"?"`

File: openstack/identity/v3/users/requests.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+// GetOptsBuilder allows extensions to add additional parameters to
+// the Get request.
+type GetOptsBuilder interface {
+	ToUserGetQuery() (string, error)
+}
+
+// GetOpts allows you to modify the details included in the Get request.
+type GetOpts struct{}
```
Comment (jrperritt) [inline openstack/identity/v3/users/requests.go]: What are the options here?

Comment (jtopjian) [inline openstack/identity/v3/users/requests.go]: There aren't any. This is only to satisfy the `ToUserGetQuery` method for possible extension. Should this be removed altogether?

Comment (jrperritt) [inline openstack/identity/v3/users/requests.go]: I don't think any of the other services offer this. I'm trying to think of what options could even by provided considering the ID uniquely identifies the user.

Comment (jtopjian) [inline openstack/identity/v3/users/requests.go]: I agree. I did the same for Projects, but in hindsight, I have no idea why. I _must_ have had a reason or known of some customization in the wild.

I'm OK with removing this and adding it later if it comes up.

**Comment by jrperritt** [comment]
This one looks good to me. We'll probably want to remove the `GetOptionsBuilder` from `v3/projects` in a separate PR.

**Comment by jtopjian** [comment]
Thanks!

I'll submit a PR for `v3/projects` now while it's on my mind.

**Pull request merged**